whois-commons/src/main/java/net/ripe/db/whois/common/apiKey/ApiKeyUtils.java

```diff
@@ -23,42 +24,55 @@ public class ApiKeyUtils {
     private static final Logger LOGGER = LoggerFactory.getLogger(ApiKeyUtils.class);
 
     public static final String APIKEY_KEY_ID_QUERY_PARAM = "keyId";
+
     public static boolean validateScope(final OAuthSession oAuthSession, final List<RpslObject> maintainers) {
-        if(StringUtils.isEmpty(oAuthSession.getScope())) {
+        if (StringUtil.isNullOrEmpty(oAuthSession.getScope())) {
             return true;
         }
 
-        final Optional<String> whoisScope = getWhoisScope(oAuthSession);
-        if(whoisScope.isEmpty()) {
+        final List<OAuthSession.ScopeFormatter> whoisScope = getWhoisScope(oAuthSession).stream().map(OAuthSession.ScopeFormatter::new).toList();
+        if (whoisScope.isEmpty()) {
             return true;
         }
 
-        final OAuthSession.ScopeFormatter scopeFormatter = new OAuthSession.ScopeFormatter(whoisScope.get());
-
-        if(StringUtils.isEmpty(scopeFormatter.getScopeKey()) || StringUtils.isEmpty(scopeFormatter.getScopeType()) || StringUtils.isEmpty(scopeFormatter.getAppName())) {
-            return true;
+        if (!isValidEnvironment(oAuthSession.getEnv(), whoisScope)) {
+            return false;
         }
 
-        return "whois".equalsIgnoreCase(scopeFormatter.getAppName())
-                    && ObjectType.MNTNER.getName().equalsIgnoreCase(scopeFormatter.getScopeType())
-                    && maintainers.stream().anyMatch( maintainer -> scopeFormatter.getScopeKey().equalsIgnoreCase(maintainer.getKey().toString()));
+        final Optional<OAuthSession.ScopeFormatter> mntnerScope = whoisScope.stream()
+                .filter(scopeFormatter -> OAuthSession.ScopeType.MNTNER.equals(scopeFormatter.getScopeType()))
+                .findFirst();
+
+        return mntnerScope.map(scopeFormatter -> maintainers.stream()
+                        .anyMatch(maintainer -> scopeFormatter.getScopeKey().equalsIgnoreCase(maintainer.getKey().toString())))
+                .orElse(true);
+
     }
 
-    private static Optional<String> getWhoisScope(OAuthSession oAuthSession) {
+    private static boolean isValidEnvironment(final Environment environment, final List<OAuthSession.ScopeFormatter> whoisScope) {
+        final Optional<OAuthSession.ScopeFormatter> environmentScope = whoisScope.stream()
+                .filter(scopeFormatter -> OAuthSession.ScopeType.ENVIRONMENT.equals(scopeFormatter.getScopeType()))
+                .findFirst();
+
+        return environmentScope.map(scopeFormatter -> environment.name().equals(scopeFormatter.getScopeKey())).orElse(true);
+    }
+
+
+    private static List<String> getWhoisScope(final OAuthSession oAuthSession) {
```

Comment:
Can you change it to getWhoisMntnrScope ? and create new method to getWhoisEnvScope? Or use stream and groupBy scopeType?

Reply:
no, this method is for getting all the scopes starting by whois.
This is ment for removing unwanted scopes. The name is also OK